ASMDocumentation.md

```diff
@@ -0,0 +1,5841 @@
++        mov     DWORD PTR _exponent$[ebp], ecx
++        jns     SHORT $LN2@fast_float
++        mov     edx, DWORD PTR _sign$[ebp]
++        mov     DWORD PTR _temp$[ebp], edx
++$LN2@fast_float:
++        mov     eax, DWORD PTR _temp$[ebp]
++        and     eax, DWORD PTR _f$[ebp]
++        mov     DWORD PTR _temp$[ebp], eax
++        fld     DWORD PTR _temp$[ebp]
++        mov     esp, ebp
+         pop     ebp
+         ret     0
+ float fast_float_trunc(float) ENDP                      ; fast_float_trunc
+```
+
+</details>
+
+</details>
+
+---
+
+<details>
+<summary>Generals/Code/Libraries/Source/WWVegas/WWLib/cpudetect.cpp</summary>
+
+This file includes the following assembly clocks:
+
+<details>
+<summary>ASM_RDTSC</summary>
+
+```c++
+#define ASM_RDTSC _asm _emit 0x0f _asm _emit 0x31
+```
+
+This code just emits the RDTSC instruction, which is a two byte instruction, `0x0F31`.
+
+This is identical to:
+
+```c++
+#define ASM_RDTSC RDTSC
+```
+
+And it is identical, in Win32; to:
+
+```c++
+#include <intrin.h>
+
+#define ASM_RDTSC __rdtsc();
+```
+
+This may have been done for obfuscation? Some parts of the code do use the `RDTSC` instruction directly and other parts
+use the bytecodes instead, it's not exactly consistent.
+
+</details>
+
+<details>
+<summary>Calculate_Processor_Speed</summary>
+
+> **NOTE**: This code has been very slightly modified to allow it to work on modern MSVC inline assembly.
+
+```c++
+static unsigned Calculate_Processor_Speed(__int64& ticks_per_second)
+{
+	struct {
+		unsigned timer0_h;
+		unsigned timer0_l;
+		unsigned timer1_h;
+		unsigned timer1_l;
+	} Time;
+
+#ifdef WIN32
+   __asm {
+      ASM_RDTSC1:
+      mov Time.timer0_h, eax
+      mov Time.timer0_l, edx
+   }
+#elif defined(_UNIX)
+      __asm__("rdtsc");
+      __asm__("mov %eax, __Time.timer1_h");
+      __asm__("mov %edx, __Time.timer1_l");
+#endif
+
+	unsigned start=TIMEGETTIME();
+	unsigned elapsed;
+	while ((elapsed=TIMEGETTIME()-start)<200) {
+#ifdef WIN32
+      __asm {
+         ASM_RDTSC2:
+         mov Time.timer1_h, eax
+         mov Time.timer1_l, edx
+      }
+#elif defined(_UNIX)
+      __asm__ ("rdtsc");
+      __asm__("mov %eax, __Time.timer1_h");
+      __asm__("mov %edx, __Time.timer1_l");
+#endif
+	}
+
+	__int64 t=*(__int64*)&Time.timer1_h-*(__int64*)&Time.timer0_h;
+	ticks_per_second=(__int64)((1000.0/(double)elapsed)*(double)t);	// Ticks per second
+	return unsigned((double)t/(double)(elapsed*1000));
+}
+```
+
+This is uses cross-platform inline assembly to calculate processor speed.
+
+My goto equivalent is:
+
+```c++
+#ifdef _WIN32
+#include <windows.h>
+#include <intrin.h>
+#elif defined(_UNIX)
+#include <ia32intrin.h>
+#endif
+
+#define TIMEGETTIME timeGetTime
+
+static unsigned Calculate_Processor_Speed(__int64& ticks_per_second)
+{
+	struct {
+		unsigned timer0_h;
+		unsigned timer0_l;
+		unsigned timer1_h;
+		unsigned timer1_l;
+	} Time;
+
+#ifdef _WIN32
+    unsigned __int64 timeStampCounter = __rdtsc();
```

Comment:
Not sure why this code is ifdef for windows, the __rdtsc intrinsic isn't available on VC6 and modern VC supports long long.